components/places/tests/fennec_bookmarks.rs

```diff
@@ -367,6 +374,163 @@ fn test_import() -> Result<()> {
     Ok(())
 }
 
+#[test]
+fn test_timestamp_sanitization() -> Result<()> {
+    use places::api::places_api::ConnectionType;
+    use places::import::common::NOW;
+    use places::storage::bookmarks::public_node::fetch_bookmark;
+    use std::time::{Duration, SystemTime};
+
+    fn get_actual_timestamps(
+        created: Timestamp,
+        modified: Timestamp,
+    ) -> Result<(Timestamp, Timestamp)> {
+        let tmpdir = tempdir().unwrap();
+        let fennec_path = tmpdir.path().join("browser.db");
+        let fennec_db = empty_fennec_db(&fennec_path)?;
+
+        let bookmarks = [FennecBookmark {
+            _id: 6,
+            guid: "bookmarkAAAA".into(),
+            created: Some(created),
+            modified: Some(modified),
+            parent: 5,
+            url: Some("http://example.com".to_owned()),
+            ..Default::default()
+        }];
+        insert_bookmarks(&fennec_db, &get_fennec_roots())?;
+        insert_bookmarks(&fennec_db, &bookmarks)?;
+
+        let places_api = PlacesApi::new(tmpdir.path().join("places.sqlite"))?;
+        places::import::import_fennec_bookmarks(&places_api, fennec_path)?;
+
+        let reader = places_api.open_connection(ConnectionType::ReadOnly)?;
+        let b = fetch_bookmark(&reader, &Guid::from("bookmarkAAAA"), true)?.unwrap();
+        // regardless of what our caller asserts, modified must never be earlier than created.
+        assert!(b.last_modified >= b.date_added);
+        Ok((b.date_added, b.last_modified))
+    }
+
+    let now = *NOW;
+    let stnow: SystemTime = now.into();
+    let earlier: Timestamp = (stnow - Duration::new(10, 0)).into();
+    let later: Timestamp = (stnow + Duration::new(10000, 0)).into();
+    println!("Timestamp tests have now as {:?}", now);
+
+    // sane timestamps, times equal -> as specified
+    assert_eq!(get_actual_timestamps(now, now)?, (now, now));
+    assert_eq!(get_actual_timestamps(earlier, earlier)?, (earlier, earlier));
+    // sane timestamps, modified later than created -> as specified
+    assert_eq!(get_actual_timestamps(earlier, now)?, (earlier, now));
+
+    // sane timestamps, modified earlier than created -> both set to created.
+    assert_eq!(get_actual_timestamps(now, earlier)?, (now, now));
+
+    // NOTE: The results of testing not-sane is somewhat arbitrary - there are
+    // a number of results which would be fine - so long as the 'created can't
+    // be after modified' invalirant holds true (which is checked in
+    // get_actual_timestamps())
+
+    // created in the past and sane, modified not sane (too early) -> created as specified, modified = now
+    // (easily argued that modified -> now is better, but that's tricky in the sql)
```

Comment:
Nit: Did you mean `modified -> created` here? (And I agree, it might be better, but tricky and we're already resetting it anyway, so leaving it as `now` is fine!)